api/clients/accountant.go

```diff
@@ -0,0 +1,169 @@
+	return &a
+}
+
+// BlobPaymentInfo calculates and records payment information. The accountant
+// will attempt to use the active reservation first and check for quorum settings,
+// then on-demand if the reservation is not available. The returned values are
+// bin index for reservation payments and cumulative payment for on-demand payments,
+// and both fields are used to create the payment header and signature
+func (a *Accountant) BlobPaymentInfo(ctx context.Context, numSymbols uint64, quorumNumbers []uint8) (uint32, *big.Int, error) {
+	now := time.Now().Unix()
+	currentBinIndex := meterer.GetBinIndex(uint64(now), a.reservationWindow)
+
+	a.usageLock.Lock()
+	defer a.usageLock.Unlock()
+	relativeBinRecord := a.GetRelativeBinRecord(currentBinIndex)
+	relativeBinRecord.Usage += numSymbols
+
+	// first attempt to use the active reservation
+	binLimit := a.reservation.SymbolsPerSec * uint64(a.reservationWindow)
+	if relativeBinRecord.Usage <= binLimit {
+		if err := QuorumCheck(quorumNumbers, a.reservation.QuorumNumbers); err != nil {
+			return 0, big.NewInt(0), err
+		}
+		return currentBinIndex, big.NewInt(0), nil
+	}
+
+	overflowBinRecord := a.GetRelativeBinRecord(currentBinIndex + 2)
+	// Allow one overflow when the overflow bin is empty, the current usage and new length are both less than the limit
+	if overflowBinRecord.Usage == 0 && relativeBinRecord.Usage-numSymbols < binLimit && numSymbols <= binLimit {
+		overflowBinRecord.Usage += relativeBinRecord.Usage - binLimit
+		if err := QuorumCheck(quorumNumbers, a.reservation.QuorumNumbers); err != nil {
+			return 0, big.NewInt(0), err
+		}
+		return currentBinIndex, big.NewInt(0), nil
+	}
+
+	// reservation not available, attempt on-demand
+	//todo: rollback later if disperser respond with some type of rejection?
+	relativeBinRecord.Usage -= numSymbols
+	incrementRequired := big.NewInt(int64(a.PaymentCharged(uint(numSymbols))))
+	a.cumulativePayment.Add(a.cumulativePayment, incrementRequired)
+	if a.cumulativePayment.Cmp(a.onDemand.CumulativePayment) <= 0 {
+		if err := QuorumCheck(quorumNumbers, []uint8{0, 1}); err != nil {
+			return 0, big.NewInt(0), err
+		}
+		return 0, a.cumulativePayment, nil
+	}
+	return 0, big.NewInt(0), fmt.Errorf("neither reservation nor on-demand payment is available")
+}
+
+// accountant provides and records payment information
+func (a *Accountant) AccountBlob(ctx context.Context, numSymbols uint64, quorums []uint8) (*commonpb.PaymentHeader, []byte, error) {
+	binIndex, cumulativePayment, err := a.BlobPaymentInfo(ctx, numSymbols, quorums)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	accountID := a.paymentSigner.GetAccountID()
+	pm := &core.PaymentMetadata{
+		AccountID:         accountID,
+		BinIndex:          binIndex,
+		CumulativePayment: cumulativePayment,
+	}
+	protoPaymentHeader := pm.ConvertToProtoPaymentHeader()
+
+	signature, err := a.paymentSigner.SignBlobPayment(protoPaymentHeader)
```

Comment:
uhhhhh I think it was because `DispersePaidBlobRequest` gave me the proto struct. This is a good question, I thought about it a bit and added a conversion function (proto->golang) so we are now using the local struct